napari/utils/events/containers/_nested_list.py

```diff
@@ -258,76 +265,121 @@ def _non_negative_index(
+                _idx: list[Index] = list(idx)
+                if isinstance(_idx[len(dest_par)], slice):
+                    raise NotImplementedError(
+                        "Can't yet deal with slice source indices in multimove"
+                    )
+                _idx[len(dest_par)] += sum(map(lambda x: x <= z, dumped))  # type: ignore
+                idx = tuple(_idx)
+
             src_par, src_i = split_nested_index(idx)
             if isinstance(src_i, slice):
                 raise ValueError("Terminal source index may not be a slice")
-            _store.append(self[idx])
-            # we need to decrement the destination index by 1 for each time we
-            # pull items in front of dest_i from the same parent as the dest
-            if src_par == dest_par and src_i < dest_i:
-                shift_dest -= 1
+            src_i = cast(int, src_i)
+            if src_i < 0:
+                src_i += len(self[src_par])
 
-        # TODO: add the appropriate moving/moved events
-        with self.events.blocker_all():
-            # delete the stored items from the list
-            for idx in sorted(sources, reverse=True):
-                del self[idx]
-            dest_i += shift_dest
-            # insert into the destination
-            self[dest_par][dest_i:dest_i] = _store
+            # we need to decrement the src_i by 1 for each time we have
+            # previously pulled items out from in front of the src_i
+            src_i -= sum(map(lambda x: x <= src_i, popped.get(src_par, [])))
+
+            # we need to decrement the dest_i by 1 for each time we have
+            # previously pulled items out from in front of the dest_i
+            ddec = sum(map(lambda x: x <= dest_i, popped.get(dest_par, [])))
+
+            # skip noop
+            if src_par == dest_par and src_i == dest_i - ddec:
+                continue
+
+            yield src_par + (src_i,), dest_par + (dest_i - ddec,)
+            popped[src_par].append(src_i)
+            dumped.append(dest_i - ddec)
+
+    def move_multiple(
+        self, sources: Iterable[NestedIndex], dest_index: NestedIndex
+    ) -> int:
+        """Move a batch of nested indices, to a single destination.
+
+        Parameters
+        ----------
+        sources : Iterable[tuple[int, ...ƒ]]
+            An iterable of tuple[int] that should be moved to ``dest_index``.
+            (Note: currently, the order of ``sources`` will NOT be maintained.)
+        dest_index : Tuple[int]
+            The destination for sources.
+        """
+        logger.debug(
+            f"move_multiple(sources={sources}, dest_index={dest_index})"
+        )
+
+        # calling list here makes sure that there are no index errors up front
+        move_plan = list(self._nested_move_plan(sources, dest_index))
+
+        # more complicated when moving multiple objects.
+        # don't assume index adjacency ... so move one at a time
+        with self.events.reordered.blocker():
```

Comment:
Yeah, I remember that.  I'm _hoping_ this will be ok.  This is the thing that took days to figure out.  to keep a very long story somewhat short: there are basically two ways to sync movements between a python model and a Qt Model
1) using [`begin/endMoveRows`](https://doc.qt.io/qt-5/qabstractitemmodel.html#beginMoveRows), (which only supports moving contiguous items)
2) using [`layoutAboutToBeChanged`](https://doc.qt.io/qt-5/qabstractitemmodel.html#layoutAboutToBeChanged) and manually using [`changePersistentIndex`](https://doc.qt.io/qt-5/qabstractitemmodel.html#changePersistentIndex) to update all of the QModel indices.

the former is a "higher level" API that takes care of some super annoying stuff for you, but it doesn't work with non-contiguous movements (which will be common with drag/drop).  The latter is tremendously flexible, but quite complicated, because you not only need to update the moving indices, but all of the other indices in the model that are "affected" by the moving items.

So, after trying to use the lower level API, I backed off and settled on the `beginMoveRows` API... which basically requires me to move these items one at a time, emitting events each time.  Hopefully it won't break whatever we were avoiding in #2087?

Reply:
Yeah, I can imagine the complexity here. Let's press on, but thanks for those additional links, might come in handy in the future!!